File: web/src/canvas/nodes/eventFilter.ts
import type {
  ChannelDescriptor,
  EventFacet,
  EventFilterNode,
  EventKindFacets,
  FilterMode,
} from "../../lib/types";

export const MAX_FILTER_IDS = 256;
export const MAX_FILTER_DURATION_MS = 600_000;
export const MAX_FILTER_TEXT_LEN = 128;

export type TriState = "any" | "yes" | "no";

export function parseIds(text: string): number[] {
  const seen = new Set<number>();
  for (const token of text.split(/[\s,]+/)) {
    if (token === "") {
      continue;
    }
    const value = Number(token);
    if (!Number.isInteger(value) || value < 0) {
      continue;
    }
    seen.add(value);
  }
  return [...seen].slice(0, MAX_FILTER_IDS);
}

export function formatIds(ids: readonly number[] | undefined): string {
  return (ids ?? []).join(", ");
}

export function parseWords(text: string): string[] {
  const seen = new Set<string>();
  for (const token of text.split(/[\s,]+/)) {
    const word = token.trim();
    if (word !== "" && word.length <= MAX_FILTER_TEXT_LEN) {
      seen.add(word);
    }
  }
  return [...seen].slice(0, MAX_FILTER_IDS);
}

export function formatWords(words: readonly string[] | undefined): string {
  return (words ?? []).join(", ");
}

export function facetsOf(kind: string, facets: readonly EventKindFacets[]): readonly EventFacet[] {
  return facets.find((entry) => entry.kind === kind)?.facets ?? [];
}

function everyKindHas(
  kinds: readonly string[],
  facet: EventFacet,
  facets: readonly EventKindFacets[],
): boolean {
  return kinds.length > 0 && kinds.every((kind) => facetsOf(kind, facets).includes(facet));
}

export function stationLabel(kinds: readonly string[], facets: readonly EventKindFacets[]): string {
  if (kinds.length > 0 && kinds.every((kind) => kind === "adsb")) {
    return "Aircraft";
  }
  if (kinds.length > 0 && kinds.every((kind) => kind === "ais")) {
    return "Vessels";
  }
  if (everyKindHas(kinds, "voice", facets)) {
    return "Radios seen";
  }
  return "Stations";
}

export function toTriState(value: boolean | null | undefined): TriState {
  if (value == null) {
    return "any";
  }
  return value ? "yes" : "no";
}

export function fromTriState(state: TriState): boolean | undefined {
  if (state === "any") {
    return undefined;
  }
  return state === "yes";
}

export interface WiredSource {
  channelType?: string;
  recordsCalls: boolean;
  trunk: boolean;
  monitor?: boolean;
}

export function kindsOffered(
  sources: readonly WiredSource[],
  descriptors: readonly ChannelDescriptor[],
): string[] {
  const kinds = new Set<string>();
  for (const source of sources) {
    if (source.monitor) {
      kinds.add("transmission");
      for (const descriptor of descriptors) {
        if (descriptor.decoder_kind != null) kinds.add(descriptor.decoder_kind);
      }
      kinds.add("broadcast_data");
    }
    if (source.trunk) {
      kinds.add("dv");
    }
    const kind = descriptors.find((d) => d.type_id === source.channelType)?.decoder_kind;
    if (kind != null) {
      kinds.add(kind);
      if (kind === "broadcast") kinds.add("broadcast_data");
    }
    if (source.recordsCalls) {
      kinds.add("call");
    }
  }
  return [...kinds].toSorted();
}

export const FILTER_MODES: readonly { value: FilterMode; label: string; title: string }[] = [
  { value: "keep", label: "Keep", title: "Only events matching every rule pass" },
  { value: "drop", label: "Drop", title: "Events matching every rule are removed" },
];

export function filterMode(filter: EventFilterNode): FilterMode {
  return filter.mode ?? "keep";
}

export function filterSaid(filter: EventFilterNode): string {
  const parts: string[] = [];
  const kinds = filter.kinds ?? [];
  parts.push(kinds.length === 0 ? "every event" : kinds.join(", "));
  if ((filter.stations ?? []).length > 0) {
    parts.push(formatWords(filter.stations));
  }
  if ((filter.contains ?? "") !== "") {
    parts.push(`"${filter.contains}"`);
  }
  if (filter.has_position != null) {
    parts.push(filter.has_position ? "with a fix" : "without a fix");
  }
  if ((filter.talkgroups ?? []).length > 0) {
    parts.push(`TG ${formatIds(filter.talkgroups)}`);
  }
  if ((filter.radios ?? []).length > 0) {
    parts.push(`radio ${formatIds(filter.radios)}`);
  }
  if (filter.encrypted != null) {
    parts.push(filter.encrypted ? "encrypted" : "clear");
  }
  if (filter.emergency != null) {
    parts.push(filter.emergency ? "emergency" : "routine");
  }
  if ((filter.min_duration_ms ?? 0) > 0) {
    parts.push(`over ${((filter.min_duration_ms ?? 0) / 1000).toFixed(1)} s`);
  }
  if (filterMode(filter) === "drop" && parts.length === 1 && kinds.length === 0) {
    return "drop nothing";
  }
  return `${filterMode(filter)} ${parts.join(" · ")}`;
}

export type PredicateKey =
  | "stations"
  | "contains"
  | "has_position"
  | "talkgroups"
  | "radios"
  | "encrypted"
  | "emergency"
  | "min_duration_ms";

export function predicatesFor(
  kinds: readonly string[],
  facets: readonly EventKindFacets[],
): PredicateKey[] {
  const touches = (facet: EventFacet) =>
    kinds.length === 0 || kinds.some((kind) => facetsOf(kind, facets).includes(facet));
  const shown: PredicateKey[] = ["stations", "contains"];
  if (touches("position")) {
    shown.push("has_position");
  }
  if (touches("voice")) {
    shown.push("talkgroups", "radios", "encrypted", "emergency");
  }
  if (touches("duration")) {
    shown.push("min_duration_ms");
  }
  return shown;
}

export interface PredicateSection {
  key: string;
  title: string;
  applies: string[];
  predicates: PredicateKey[];
}

export function sectionsFor(
  kinds: readonly string[],
  facets: readonly EventKindFacets[],
): PredicateSection[] {
  const shown = predicatesFor(kinds, facets);
  const pick = (keys: PredicateKey[]) => keys.filter((key) => shown.includes(key));
  const scope = (facet: EventFacet) =>
    kinds.filter((kind) => facetsOf(kind, facets).includes(facet)).toSorted();
  const sections: PredicateSection[] = [];
  const any = pick(["stations", "contains"]);
  if (any.length > 0) {
    sections.push({
      key: "any",
      title: "Any event",
      applies: [...kinds].toSorted(),
      predicates: any,
    });
  }
  const position = pick(["has_position"]);
  if (position.length > 0) {
    sections.push({
      key: "position",
      title: "Position",
      applies: scope("position"),
      predicates: position,
    });
  }
  const voice = pick(["talkgroups", "radios", "encrypted", "emergency", "min_duration_ms"]);
  if (voice.length > 0) {
    sections.push({
      key: "voice",
      title: "Voice",
      applies: scope("voice"),
      predicates: voice,
    });
  }
  return sections;
}

File: apps/native/src/ui/faces/event_filter.rs
use sdrmm_wire::{
    filter::{EventFilterNode, FilterMode, MAX_FILTER_DURATION_MS},
    patch::NodeBody,
};
use zgui::prelude::*;

use crate::{
    decoders::{
        filter::{
            Predicate, TriState, filter_said, format_ids, kinds_offered, parse_ids, parse_words,
            sections_for, station_label,
        },
        log::kind_label,
        wiring::{event_sources_of, wired_sources_of},
    },
    store::Store,
    ui::{
        kit_decoders,
        params::entry,
        widgets::{check, pick, row_field, segments},
    },
};

type Edit = std::rc::Rc<dyn Fn(Box<dyn FnOnce(&mut EventFilterNode)>)>;

fn settings_of(store: Store, node: &str) -> EventFilterNode {
    match store.graph.get().node(node).map(|found| &found.body) {
        Some(NodeBody::EventFilter(filter)) => filter.clone(),
        _ => EventFilterNode::default(),
    }
}

pub fn face(store: Store, node: String) -> impl IntoView {
    kit_decoders::install();
    let settings = {
        let node = node.clone();
        Memo::new(move |_| settings_of(store, &node))
    };
    let wired = {
        let node = node.clone();
        Memo::new(move |_| !event_sources_of(&store.graph.get(), &node).is_empty())
    };
    let offered = {
        let node = node.clone();
        Memo::new(move |_| {
            kinds_offered(
                &wired_sources_of(&store.graph.get(), &node),
                &store.channel_types.get(),
            )
        })
    };
    let edit: Edit = std::rc::Rc::new(move |change| {
        let node = node.clone();
        store.edit_graph(move |graph| {
            if let Some(found) = graph.nodes.iter_mut().find(|found| found.id == node)
                && let NodeBody::EventFilter(filter) = &mut found.body
            {
                change(filter);
            }
        });
    });
    let hint = move || {
        let said = if !wired.get() {
            "Wire decoder events in".to_owned()
        } else if offered.with(Vec::is_empty) {
            "Nothing wired in emits events".to_owned()
        } else {
            filter_said(&settings.get())
        };
        view! { text(class = "hint") {{said}} }
    };
    let mode = Signal::derive(move || settings.get().mode);
    let set_mode = {
        let edit = edit.clone();
        move |mode: FilterMode| edit(Box::new(move |filter| filter.mode = mode))
    };
    let kinds = {
        let edit = edit.clone();
        move || kind_chips(settings, offered, edit.clone())
    };
    let sections = move || {
        let narrowed = settings.with(|s| s.kinds.clone());
        let narrowed = if narrowed.is_empty() {
            offered.get()
        } else {
            narrowed
        };
        let current = settings.get_untracked();
        sections_for(&narrowed)
            .into_iter()
            .map(|section| {
                let applies = section.applies.join(" · ");
                let rows: Vec<AnyView> = section
                    .predicates
                    .iter()
                    .map(|which| predicate(*which, &current, &narrowed, edit.clone()))
                    .collect();
                view! {
                    column(class = "dk-pane") {
                        row(class = "dk-line") {
                            text(class = "legend") {{section.title}}
                            text(class = "dk-dim") {{applies}}
                        }
                        {rows}
                    }
                }
            })
            .collect::<Vec<_>>()
    };
    view! {
        column(class = "face", {..kit_decoders::no_pan()}) {
            {hint}
            {row_field("Mode", segments(vec![(FilterMode::Keep, "Keep"), (FilterMode::Drop, "Drop")], mode, set_mode))}
            {kinds}
            {sections}
        }
    }
}

fn kind_chips(
    settings: Memo<EventFilterNode>,
    offered: Memo<Vec<String>>,
    edit: Edit,
) -> Option<AnyView> {
    let offered = offered.get();
    if offered.len() < 2 {
        return None;
    }
    let chips: Vec<_> = offered
        .into_iter()
        .map(|kind| {
            let held = kind.clone();
            let on = Signal::derive(move || settings.with(|s| s.kinds.contains(&held)));
            let edit = edit.clone();
            let label = kind_label(&kind);
            let toggle = move |pressed: bool| {
                let kind = kind.clone();
                edit(Box::new(move |filter| {
                    filter.kinds.retain(|held| *held != kind);
                    if pressed {
                        filter.kinds.push(kind);
                        filter.kinds.sort();
                    }
                }));
            };
            view! {
                row(class = "dk-line dk-chip") {
                    {check(on, toggle)}
                    text {{label}}
                }
            }
        })
        .collect();
    Some(AnyView::new(view! {
        column(class = "dk-pane") {
            text(class = "legend") {"Kinds"}
            row(class = "dk-line") {{chips}}
        }
    }))
}

fn words(
    name: &'static str,
    value: String,
    edit: Edit,
    apply: fn(&mut EventFilterNode, String),
) -> AnyView {
    AnyView::new(row_field(
        name,
        entry(Signal::stored(value), name.to_owned(), false, move |text| {
            edit(Box::new(move |filter| apply(filter, text)));
            Ok(())
        }),
    ))
}

fn tri(
    name: &'static str,
    value: Option<bool>,
    edit: Edit,
    apply: fn(&mut EventFilterNode, Option<bool>),
) -> AnyView {
    let options = vec![
        (TriState::Any, "Either".to_owned()),
        (TriState::Yes, "Only".to_owned()),
        (TriState::No, "Never".to_owned()),
    ];
    let chosen = Signal::stored(Some(TriState::of(value)));
    AnyView::new(row_field(
        name,
        pick(options, chosen, move |state: TriState| {
            edit(Box::new(move |filter| apply(filter, state.value())));
        }),
    ))
}

fn predicate(which: Predicate, current: &EventFilterNode, kinds: &[String], edit: Edit) -> AnyView {
    match which {
        Predicate::Stations => words(
            station_label(kinds),
            current.stations.join(", "),
            edit,
            |f, t| {
                f.stations = parse_words(&t);
            },
        ),
        Predicate::Contains => words(
            "Contains",
            current.contains.clone().unwrap_or_default(),
            edit,
            |f, t| {
                let t = t.trim();
                f.contains = (!t.is_empty()).then(|| t.to_owned());
            },
        ),
        Predicate::Talkgroups => words(
            "Talkgroups",
            format_ids(&current.talkgroups),
            edit,
            |f, t| {
                f.talkgroups = parse_ids(&t);
            },
        ),
        Predicate::Radios => words("Radios", format_ids(&current.radios), edit, |f, t| {
            f.radios = parse_ids(&t);
        }),
        Predicate::HasPosition => tri("Has position", current.has_position, edit, |f, v| {
            f.has_position = v
        }),
        Predicate::Encrypted => tri("Encrypted", current.encrypted, edit, |f, v| f.encrypted = v),
        Predicate::Emergency => tri("Emergency", current.emergency, edit, |f, v| f.emergency = v),
        Predicate::MinDuration => duration(current.min_duration_ms, edit),
    }
}

fn duration(ms: u32, edit: Edit) -> AnyView {
    let shown = format!("{:.1}", f64::from(ms) / 1000.0);
    let field = entry(
        Signal::stored(shown),
        "Longer than, seconds".to_owned(),
        false,
        move |text| {
            let seconds = text
                .trim()
                .parse::<f64>()
                .ok()
                .filter(|s| s.is_finite() && *s >= 0.0)
                .ok_or_else(|| "seconds".to_owned())?;
            let ms = ((seconds * 1000.0).round() as u32).min(MAX_FILTER_DURATION_MS);
            edit(Box::new(move |filter| filter.min_duration_ms = ms));
            Ok(())
        },
    );
    AnyView::new(row_field(
        "Longer than",
        view! { row(class = "dk-line") { {field} text(class = "dk-dim") {"s"} } },
    ))
}
